core/src/main/java/org/mule/runtime/core/internal/routing/ForeachRouter.java

```diff
@@ -0,0 +1,340 @@
+    return partEvent;
+  }
+
+  private ForeachContext createForeachContext(CoreEvent event) {
+    // Keep reference to existing rootMessage/count variables in order to restore later to support foreach nesting.
+    Object previousCounterVar = event.getVariables().containsKey(owner.getCounterVariableName())
+        ? event.getVariables().get(owner.getCounterVariableName()).getValue()
+        : null;
+    Object previousRootMessageVar =
+        event.getVariables().containsKey(owner.getRootMessageVariableName())
+            ? event.getVariables().get(owner.getRootMessageVariableName()).getValue()
+            : null;
+
+    return new ForeachContext(previousCounterVar, previousRootMessageVar, event.getMessage(), event.getItemSequenceInfo());
+  }
+
+  private Iterator<TypedValue<?>> splitRequest(CoreEvent request, String expression) {
+    Object payloadValue = request.getMessage().getPayload().getValue();
+    Iterator<TypedValue<?>> result;
+    if (DEFAULT_SPLIT_EXPRESSION.equals(expression) && payloadValue instanceof EventBuilderConfigurerList) {
+      // Support EventBuilderConfigurerList currently used by Batch Module
+      result = Iterators.transform(((EventBuilderConfigurerList<Object>) payloadValue).eventBuilderConfigurerIterator(),
+                                   TypedValue::of);
+    } else if (DEFAULT_SPLIT_EXPRESSION.equals(expression) && payloadValue instanceof EventBuilderConfigurerIterator) {
+      // Support EventBuilderConfigurerIterator currently used by Batch Module
+      result = new EventBuilderConfigurerIteratorWrapper((EventBuilderConfigurerIterator) payloadValue);
+    } else {
+      result = owner.getSplittingStrategy().split(request);
+    }
+    if (LOGGER.isDebugEnabled() && !result.hasNext()) {
+      LOGGER.debug(
+                   "<foreach> expression \"{}\" returned no results. If this is not expected please check your expression",
+                   expression);
+    }
+    return result;
+  }
+
+  /**
+   * Assembles and returns the downstream {@link Publisher <CoreEvent>}.
+   *
+   * @return the successful {@link CoreEvent} or retries exhaustion errors {@link Publisher}
+   */
+  Publisher<CoreEvent> getDownstreamPublisher() {
+    return downstreamFlux
+        .compose(downstreamPublisher -> subscriberContext()
+            .flatMapMany(downstreamContext -> downstreamPublisher.doOnSubscribe(s -> {
+              // When a transaction is active, the processing strategy executes the whole reactor chain in the same thread that
+              // performs the subscription itself. Because of this, the subscription has to be deferred until the
+              // downstreamPublisher FluxCreate#subscribe method registers the new sink in the recorder.
+              downstreamCtxReference.set(downstreamContext);
+            })));
+  }
+
+  private void subscribeUpstreamChains(Context downstreamContext) {
+    innerFlux.subscriberContext(downstreamContext).subscribe();
```

Comment:
for torubleshooting, add an error handler to subscribe so that uncaught error are at least logged.